profiling/src/profile/mod.rs

```diff
@@ -659,47 +645,60 @@ impl Profile {
             local_root_span_id_label_offset,
         };
 
-        let id = match self.samples.get_index_of(&s) {
-            None => {
-                self.samples.insert(s, values);
-                SampleId::new(self.samples.len() - 1)
+        if let Some(ts) = timestamp {
+            match self.timestamped_samples.get_mut(&s) {
+                None => {
+                    let series = vec![(ts, values)];
+                    self.timestamped_samples.insert(s, series);
+                }
+                Some(series) => series.push((ts, values)),
             }
-            Some(index) => {
-                let (_, existing_values) =
-                    self.samples.get_index_mut(index).expect("index to exist");
-                for (a, b) in existing_values.iter_mut().zip(values) {
-                    a.add_assign(b)
+        } else {
+            match self.aggregated_samples.get_mut(&s) {
+                None => {
+                    self.aggregated_samples.insert(s, values);
                 }
-                SampleId::new(index)
+                Some(v) => v.iter_mut().zip(values).for_each(|(a, b)| *a += b),
             }
         };
-
-        Ok(id)
+        Ok(())
     }
 
     /// Validates labels and converts them to the internal representation.
+    /// Extracts out the timestamp label, if it exists.
     /// Also tracks the index of the label with key "local root span id".
     fn extract_sample_labels(
         &mut self,
         sample: &api::Sample,
-    ) -> anyhow::Result<(Vec<Label>, Option<usize>)> {
+    ) -> anyhow::Result<(Vec<Label>, Option<usize>, Option<Timestamp>)> {
         let mut labels: Vec<Label> = Vec::with_capacity(sample.labels.len());
         let mut local_root_span_id_label_offset: Option<usize> = None;
+        let mut timestamp: Option<Timestamp> = None;
+
         for label in sample.labels.iter() {
             let key = self.intern(label.key);
-            let str = label
-                .str
-                .map(|s| self.intern(s))
-                .unwrap_or(StringId::zero());
-            let num_unit = label
-                .num_unit
-                .map(|s| self.intern(s))
-                .unwrap_or(StringId::zero());
+            let str = label.str.map(|s| self.intern(s));
+            let num_unit = label.num_unit.map(|s| self.intern(s));
+
+            if key == self.timestamp_key {
+                anyhow::ensure!(
+                    label.str.is_none(),
+                    "the label \"{}\" must be sent as a number, not string {}",
+                    label.str.unwrap(),
+                    label.key
+                );
+                anyhow::ensure!(label.num != 0, "the label \"{}\" must not be 0", label.key);
+                //TODO what should I ensure the units are?
+
+                timestamp = Some(NonZeroI64::new(label.num).unwrap());
+                // Don't put the timestamp into the labels
```

Comment:
Minor: May be worth expanding the comment a bit to explain that this label gets treated specially and for that reason.